// ad-validator/test/css.js
const assert = require( 'assert' );
const postcss = require( 'postcss' );
const css = require( '../' ).css;

module.exports = function () {
	it( 'it rejects syntax errors', function () {
		assert.deepEqual( css( 'html {\n\t...\n}' ), [
			{
				start: {
					index: 0,
					line: 1,
					column: 1
				},
				end: {
					index: 1,
					line: 1,
					column: 2
				},
				message: `CssSyntaxError: <css input>:2:2: Unknown word`
			}
		]);
	});

	it( 'rejects selectors without WS_DYN_INS_RAND', function () {
		assert.deepEqual( css( `div {\n\twidth:100px;\n}` ), [
			{
				start: {
					index: 0,
					line: 0,
					column: 0
				},
				end: {
					index: 3,
					line: 0,
					column: 3
				},
				message: `CSS selector 'div' is missing WS_DYN_INS_RAND`
			}
		]);
	});

	it( 'rejects @rules without WS_DYN_INS_RAND', function () {
		assert.deepEqual( css( `@keyframes foo {}` ), [
			{
				start: {
					index: 11,
					line: 0,
					column: 11
				},
				end: {
					index: 14,
					line: 0,
					column: 14
				},
				message: `CSS @rule parameter 'foo' is missing WS_DYN_INS_RAND`
			}
		]);
	});

	it( 'does not reject single quotes', function () {
		assert.deepEqual( css( `.WS_DYN_INS_RAND { content: '#'; }` ), [] );
	});
};
